apps/cowswap-frontend/src/modules/wallet/web3-react/connection/walletConnectV2.tsx

```diff
@@ -32,36 +35,87 @@ export const walletConnectV2Option = {
   id: 'wallet-connect-v2',
 }
 
-const [mainnet, ...optionalChains] = Object.keys(RPC_URLS).map(Number)
-
-const [web3WalletConnectV2, web3WalletConnectV2Hooks] = initializeConnector<AsyncConnector>(
-  (actions) =>
-    new AsyncConnector(
-      () =>
-        import('@web3-react/walletconnect-v2').then(
-          (m) =>
-            new m.WalletConnect({
-              actions,
-              options: {
-                projectId: WC_PROJECT_ID || WC_DEFAULT_PROJECT_ID,
-                chains: [mainnet],
-                optionalChains,
-                showQrModal: true,
-              },
-              onError,
-            })
-        ),
-      actions,
-      onError
-    )
-)
-
-export const walletConnectConnectionV2: Web3ReactConnection = {
-  connector: web3WalletConnectV2,
-  hooks: web3WalletConnectV2Hooks,
-  type: ConnectionType.WALLET_CONNECT_V2,
+const WC_PROJECT_ID = process.env.REACT_APP_WC_PROJECT_ID
+const WC_DEFAULT_PROJECT_ID = 'a6cc11517a10f6f12953fd67b1eb67e7'
+
+function createWalletConnectV2Connector(
+  chainId: SupportedChainId
+): [WalletConnectV2Connector, Web3ReactHooks, Web3ReactStore] {
+  return initializeConnector<WalletConnectV2Connector>(
+    (actions) =>
+      new WalletConnectV2Connector({
+        actions,
+        onError(error) {
+          console.error('WalletConnect2 ERROR:', error)
+        },
+        options: {
+          projectId: WC_PROJECT_ID || WC_DEFAULT_PROJECT_ID,
+          chains: [chainId],
+          optionalChains: ALL_SUPPORTED_CHAIN_IDS,
+          showQrModal: true,
+          rpcMap: RPC_URLS,
+        },
+      })
+  )
 }
 
+function createWc2Connection(chainId = getCurrentChainIdFromUrl()): Web3ReactConnection {
+  let [web3WalletConnectV2, web3WalletConnectV2Hooks] = createWalletConnectV2Connector(chainId)
+
+  let onActivate: (() => void) | undefined
+
+  const proxyConnector = new Proxy(
```

Comment:
Just copy-pasted the solution from https://github.com/Uniswap/interface/blob/main/src/connection/index.ts#L85

Why do we need this:
WC2 connector can be created once per network.

Let's consider the case:
1. Connect via WC2 to Mainnet
2. Disconnect
3. Try to connect to Gnosis Chain via WC2

In this case, the connection won't be established, because at step 1 the WC2 connector was created for chainId=1.
To overcome this problem we proxy WC2 connection and change it's implementation on flight on network changes.

Reply:
Nice job! 
Can you add a shorter version of this as a comment?

Reply:
Sure, added